File: backend/app/schemas/rfq_automation.py
from typing import Literal

from pydantic import BaseModel, Field

ProcurementScope = Literal[
    "pipe_supply",
    "structures_supply",
    "aggregate_supply",
    "asphalt_paving",
    "concrete_subcontract",
    "testing",
    "traffic_control",
    "landscaping",
    "earthworks_support",
    "disposal",
    "misc",
]

SourceSignal = Literal["quantity", "municipality", "estimate", "manual"]


class RFQAutomationInputItem(BaseModel):
    category: str
    description: str
    quantity: float | None = None
    unit: str | None = None
    source: SourceSignal = "manual"


class RFQAutomationRequest(BaseModel):
    project_name: str | None = None
    municipality: str | None = None
    items: list[RFQAutomationInputItem] = Field(default_factory=list)
    include_default_civil_scopes: bool = True


class RFQScopeRecommendation(BaseModel):
    scope: ProcurementScope
    title: str
    reason: str
    supplier_categories: list[str]
    required_documents: list[str]
    priority: int
    source_signals: list[SourceSignal]
    review_notes: list[str]


class RFQAutomationResponse(BaseModel):
    project_name: str | None
    municipality: str | None
    recommendation_count: int
    high_priority_count: int
    recommendations: list[RFQScopeRecommendation]
    warnings: list[str]

File: frontend/src/api/rfqAutomation.ts
export type ProcurementScope =
  | "pipe_supply"
  | "structures_supply"
  | "aggregate_supply"
  | "asphalt_paving"
  | "concrete_subcontract"
  | "testing"
  | "traffic_control"
  | "landscaping"
  | "earthworks_support"
  | "disposal"
  | "misc";

export type SourceSignal = "quantity" | "municipality" | "estimate" | "manual";

export type RFQAutomationInputItem = {
  category: string;
  description: string;
  quantity?: number | null;
  unit?: string | null;
  source: SourceSignal;
};

export type RFQAutomationRequest = {
  project_name?: string | null;
  municipality?: string | null;
  items: RFQAutomationInputItem[];
  include_default_civil_scopes: boolean;
};

export type RFQScopeRecommendation = {
  scope: ProcurementScope;
  title: string;
  reason: string;
  supplier_categories: string[];
  required_documents: string[];
  priority: number;
  source_signals: SourceSignal[];
  review_notes: string[];
};

export type RFQAutomationResponse = {
  project_name?: string | null;
  municipality?: string | null;
  recommendation_count: number;
  high_priority_count: number;
  recommendations: RFQScopeRecommendation[];
  warnings: string[];
};

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? "http://localhost:8000/api/v1";

async function request<T>(path: string, options?: RequestInit): Promise<T> {
  const response = await fetch(`${API_BASE_URL}${path}`, {
    headers: {
      "Content-Type": "application/json",
      ...options?.headers,
    },
    ...options,
  });
  if (!response.ok) throw new Error(`Request failed with ${response.status}`);
  return response.json() as Promise<T>;
}

export const rfqAutomationApi = {
  recommend: (payload: RFQAutomationRequest) =>
    request<RFQAutomationResponse>("/rfq-automation/recommend", {
      method: "POST",
      body: JSON.stringify(payload),
    }),
};
